vscode-lean4/src/infoview.ts

```diff
@@ -3,16 +3,44 @@ import {
     commands, Disposable, DocumentSelector,
     ExtensionContext, languages, Range,
     Selection, TextEditor, TextEditorRevealType,
-    Uri, ViewColumn, WebviewPanel, window, workspace, env, Position, Diagnostic,
+    Uri, ViewColumn, WebviewPanel, window, workspace, env, Position,
 } from 'vscode';
-import { EditorApi, InfoviewApi, LeanFileProgressParams, TextInsertKind } from '@lean4/infoview';
+import { EditorApi, InfoviewApi, LeanFileProgressParams, TextInsertKind, RpcConnectParams, RpcConnected, RpcKeepAliveParams } from '@lean4/infoview';
 import { LeanClient } from './leanclient';
 import { getInfoViewAllErrorsOnLine, getInfoViewAutoOpen, getInfoViewAutoOpenShowGoal,
     getInfoViewFilterIndex, getInfoViewStyle, getInfoViewTacticStateFilters } from './config';
 import { Rpc } from './rpc';
-import { PublishDiagnosticsParams } from 'vscode-languageclient';
 import * as ls from 'vscode-languageserver-protocol'
 
+const keepAlivePeriodMs = 10000
```

Comment:
How did we come up with 10 seconds?  Can it be 20 seconds?  If you can move it to 20 seconds you will double my laptop battery life...

Reply:
The 10s was chosen arbitrarily, to be more or less at an order of magnitude negligible in terms of power usage but reasonably short to a human. On which point, I am assuming you are joking about the battery life? A short computation every few seconds will have no effect at all at power usage in a normal editing session.

Reply:
Actually I wasn't joking, if there's nothing else going on on my machine, switching from 10 to 20 seconds would literally double my battery life.  of course, that's a big if.  Every other app running on my system would have to have an excellently silent idle loop, and my RAM would have to be consuming very low amount of power (which is also not true).  So it's sort of a joke, but it's more a question of being a "well behaved" application that doesn't needlessly suck power while idle.

Reply:
I see, apologies for assuming that then. Still, looking at any modern system in its idle state, one observes multiple background services such as the window manager, OS interrupt handlers, sync services, a browser if you have one open, etc, waking up and doing bits of work multiple times a second. I am really confident that this 10s vs 20s does not matter for power draw at all. So changing this value will not make us any better behaved. Please disregard the humans comment (bad phrasing) and just take 10s to be the arbitrarily chosen value it is, within an order of magnitude of [TCP keepalives](https://en.wikipedia.org/wiki/Keepalive#TCP_keepalive) which inspired the protocol.

Reply:
Ok, I'll stop pushing then, and defer to my other dumb question below about whether we need keepalive at all.

Reply:
Another related effect of keep-alives is that they keep the lean process in memory, which is undesirable in low-memory situations (where you want the OS to swap out the process).

Also note that we're not talking about a single process here.  It's extension host + lean watchdog + lean file worker (+ browser tab if we'd do it in the infoview).  At the moment the infoview also keeps sessions connected to all opened files, whether they are used or not (though we could conceivable optimize that), so effectively it's 10 * lean file worker.  And all of that * number of open vscode windows.

Generating the keep-alives in the infoview is also undesirable if you use vscode in a browser, as it will cause constant network traffic.  (I believe the extension host is run remotely.)

> Still, looking at any modern system in its idle state, one observes multiple background services such as the window manager, OS interrupt handlers, sync services, a browser if you have one open, etc, waking up and doing bits of work multiple times a second.

A couple of years ago there was a huge push to reduce wakups in the linux ecosystem (around the time Intel introduced tools like `powertop` with the explicit purpose of finding misbehaving programs that wake up unnecessarily).  Anecdotally, this effort has led to much improved battery runtimes (with 10+ hours being the expected norm instead of the exception).  Of course there are still some buggy programs out there, but I don't want to contribute to the problem if we can easily avoid it.